drivers/i2c/i2c_nrfx_twi_common.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2024, Croxel Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <nrfx_twi.h>
+#include "i2c_nrfx_twi_common.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(i2c_nrfx_twi);
+
+int i2c_nrfx_twi_init(const struct device *dev)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	nrfx_err_t result = nrfx_twi_init(&config->twi, &config->config,
+					  config->event_handler, (void *)dev);
+	if (result != NRFX_SUCCESS) {
+		LOG_ERR("Failed to initialize device: %s",
+			    dev->name);
+		return -EBUSY;
+	}
+
+	return 0;
+}
+
+int i2c_nrfx_twi_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	struct i2c_nrfx_twi_common_data *data = dev->data;
+	nrfx_twi_t const *inst = &config->twi;
+
+	if (I2C_ADDR_10_BITS & dev_config) {
+		return -EINVAL;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		nrf_twi_frequency_set(inst->p_twi, NRF_TWI_FREQ_100K);
+		break;
+	case I2C_SPEED_FAST:
+		nrf_twi_frequency_set(inst->p_twi, NRF_TWI_FREQ_400K);
+		break;
+	default:
+		LOG_ERR("unsupported speed");
+		return -EINVAL;
+	}
+	data->dev_config = dev_config;
+
+	return 0;
+}
+
+int i2c_nrfx_twi_recover_bus(const struct device *dev)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	uint32_t scl_pin;
+	uint32_t sda_pin;
+	nrfx_err_t err;
+
+	scl_pin = nrf_twi_scl_pin_get(config->twi.p_twi);
+	sda_pin = nrf_twi_sda_pin_get(config->twi.p_twi);
+
+	err = nrfx_twi_bus_recover(scl_pin, sda_pin);
+	return (err == NRFX_SUCCESS ? 0 : -EBUSY);
+}
+
+int i2c_nrfx_twi_msg_transfer(const struct device *dev, uint8_t flags,
+			      uint8_t *buf, size_t buf_len, uint16_t i2c_addr)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	int ret = 0;
+	uint32_t xfer_flags = 0;
+	nrfx_err_t res;
+	nrfx_twi_xfer_desc_t cur_xfer = {
+		.p_primary_buf = buf,
+		.primary_length = buf_len,
+		.address = i2c_addr,
+		.type = (flags & I2C_MSG_READ) ?
+			 NRFX_TWI_XFER_RX : NRFX_TWI_XFER_TX,
+	};
+
+	if (flags & I2C_MSG_ADDR_10_BITS) {
+		LOG_ERR("10-bit I2C Addr devices not supported");
+		ret = -ENOTSUP;
+	} else if (!(flags & I2C_MSG_STOP)) {
```

Comment:
This part is missing:
```
			/* - if the transfer consists of more messages
			 *   and the I2C repeated START is not requested
			 *   to appear before the next message, suspend
			 *   the transfer after the current message,
			 *   so that it can be resumed with the next one,
			 *   resulting in the two messages merged into
			 *   a continuous transfer on the bus
			 */
			if ((i < (num_msgs - 1)) &&
			    !(msgs[i + 1].flags & I2C_MSG_RESTART)) {
				xfer_flags |= NRFX_TWI_FLAG_SUSPEND;
                        }
```

Reply:
That's true. Thanks for pointing it out. I was struggling to refactor this into common code.

I have now resolved to add a flag `bool more_msgs` to represent this condition in `i2c_nrfx_twi_msg_transfer`.